### ISSUE TITLE
Abstract methods in ActionScript

### ISSUE THREAD

**Issue by joshtynjala** [apache/royale-compiler#75]
Followup to #73: [Abstract Classes in ActionScript](https://github.com/apache/royale-compiler/issues/73).

## Proposed syntax

``` actionscript
package com.example
{
	public abstract class MyAbstractClass
	{
		public MyAbstractClass()
		{
			
		}

		public abstract function myAbstractFunction(arg0:Number):String;
	}
}
```

``` actionscript
package com.example
{
	public class MyConcreteClass extends MyAbstractClass
	{
		public MyConcreteClass ()
		{
			
		}

		override public function myAbstractFunction(arg0:Number):String
		{
			return null;
		}
	}
}
```

## Semantics

Similar to the `UnimplementedInterfaceMethodProblem` that is created when a class does not implement an interface method, there should be a new problem for abstract classes with this message:

> abstract method myAbstractMethod in class MyAbstractClass not implemented by class MyConcreteClass

Similar to the `InterfaceMethodWithBodyProblem` or `NativeMethodWithBodyProblem` that is created when adding a body to methods in an interface or methods marked `native`, there should be a new problem for abstract classes with this message:

> Abstract methods must not have a body.

A class must be abstract to contain abstract methods. Marking a method as abstract on a non-abstract class should result in a problem being created with this message:

> The abstract attribute may be used only on method definitions of an abstract class.

## Bytecode

In SWF bytecode, the abstract method compiles with `[RoyaleAbstract]` metadata instead:

``` actionscript
package com.example
{
	[RoyaleAbstract]
	public class MyAbstractClass
	{
		public MyAbstractClass()
		{
			
		}

		[RoyaleAbstract]
		public function myAbstractFunction(arg0:Number):String;
	}
}
```

Testing in SWF, a method body does not seem to be required. Calling an abstract method implemented on a concrete class seems to work correctly. No bytecode verify errors.

## How to enable

New ActionScript syntax should be opt-in. Developers can enable abstract methods using the `allow-abstract-classes` compiler option (which was already created for abstract classes).

``` sh
mxmlc -allow-abstract-classes App.mxml
```

**Comment by piotrzarzycki21** [comment]
+1 for having abstract method enabled once abstract classes are enabled. 
**Another question is:** Should abstract method have to be override in class which inherits from abstract class ?

**Comment by joshtynjala** [comment]
> Another question is: Should abstract method have to be override in class which inherits from abstract class ?

I found this rule in Java, which I think makes sense:

> When an abstract class is subclassed, the subclass usually provides implementations for all of the abstract methods in its parent class. However, if it does not, then the subclass must also be declared abstract.

**Comment by carlosrovira** [comment]
Hi @joshtynjala 
can you summarize latest changes so we all can know better what implies?
thanks! :)

**Comment by joshtynjala** [comment]
@carlosrovira I have expanded abstract classes to include support for abstract methods.

If a class is marked as `abstract`, it cannot be instantiated directly with `new`. It must be subclassed.

An abstract class may define variables, properties, and methods normally, but it may also optionally define methods marked as `abstract`. An abstract method cannot be `static`, `final`, or a getter/setter, and it must not have a body. A subclass that is not abstract must provide implementations for any abstract methods, similar to how implementations must be provided for methods when implementing an interface.

**Comment by carlosrovira** [comment]
Thanks Josh, I was figuring most of what you comment, but is great that you listed the real things that comes in with this update. Great additions! thanks for working on this! :)
